keps/20200901-standard-crd.md

```diff
@@ -0,0 +1,163 @@
+status: draft
+---
+```
+
+# Standardize ExternalSecret CRD
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Terminology](#terminology)
+- [Use-Cases](#use-cases)
+- [Proposal](#proposal)
+  - [API](#api)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This is a proposal to standardize the `ExternalSecret` CRD in an combined effort through all projects that deal with syncing external secrets. This proposal aims to do find the _common denominator_ for all users of `ExternalSecrets`.
+
+## Motivation
+
+There are a lot of different projects in the wild that essentially do the same thing: sync secrets with Kubernetes. The idea is to unify efforts into a single project that serves the needs of all users in this problem space.
+
+As a starting point i would like to define a **common denominator** for a CustomResourceDefinition that serves all known use-cases. This CRD should follow the standard alpha -> beta -> GA feature process.
+
+Once the CRD API is defined we can move on with more delicate discussions about technology, organization and processes.
+
+### Goals
+
+- Define a alpha CRD
+- Fully document the Spec and use-cases
+
+### Non-Goals
+
+This KEP proposes the CRD Spec and documents the use-cases, not the choice of technology or migration path towards implementing the CRD.
+
+## Terminology
+
+* Kubernetes External Secrets `KES`: A Application that runs a control loop which syncs secrets
+* KES `instance`: A single entity that runs a control loop.
+* ExternalSecret `ES`: A CustomResource that declares which secrets should be synced
+* Backend: A **source** for secrets. The Backend is external to the cluster. E.g.: Alibabacloud SecretsManager, AWS SystemsManager, Azure KeyVault...
+* Frontend: A **sink** for the synced secrets. Usually a `ConfigMap` or `Secret`
+* Secret: credentials that act as a key to sensitive information
+
+## Use-Cases
+
+The following use cases are taken from
+
+1. AS a KES user i want to run multiple KES instances per cluster (e.g. one KES instance per DEV/PROD)
+2. AS a KES user i want to integrate **multiple backends** from a **single KES instance** (e.g. dev namespace has access only to dev secrets)
+3. AS a KES user i want to control the sink for the secrets (aka frontend: store secret as `kind=ConfigMap` or `kind=Secret`)
+4. AS a KES user i want to fetch **from multiple** Backends and store the secrets **in a single** Frontend
+5. AS a KES operator i want to limit the access to certain backends or subresources (e.g. having one central KES instance that handles all ES - similar to `iam.amazonaws.com/permitted` annotation per namespace)
+
+## Proposal
+
+### API
+
+The `ExternalSecret` CustomResourceDefinition is **namespaced**. It defines the following:
+1. source for the secret (backend)
+2. sink for the secret (fronted)
+3. and a mapping to translate the keys
+
+```yaml
+apiVersion: kes.io/v1alpha1
```

Comment:
not sure if this is a good idea maybe we should mount this under `k8s.io`, like `external-secrets.k8s.io`

Reply:
`external-secrets.k8s.io` sounds nice 😏 unless someone gets unhappy with the use of *.k8s.io

Reply:
I stumbled upon this slack thread in [sig-architecture](https://kubernetes.slack.com/archives/C5P3FE08M/p1600968242008600) which points to the [api-review process](https://github.com/kubernetes/community/blob/master/sig-architecture/api-review-process.md#voluntary). From my understanding we really shouldn't use `k8s.io`. If we plan to move the effort towards a Kubernetes SIG we should plan to use `x-k8s.io`.